nemo_aligner/models/nlp/gpt/megatron_gpt_reward_model.py

```diff
@@ -350,93 +353,63 @@ def finish_validation_step(self):
-                    tokenizer, inputs, max_seq_length, False, add_EOS=add_EOS
-                )
-            if context_length_tensor.dim() == 1:
-                context_length_tensor = context_length_tensor.unsqueeze(-1)
-
-        context_length_tensor = broadcast_2d_tensor(context_length_tensor, src=0, group=None, dtype=torch.int64)
-        if context_length_tensor.dim() == 2:
-            context_length_tensor = context_length_tensor.squeeze(-1)
-
-        context_tokens_tensor = broadcast_2d_tensor(context_tokens_tensor, src=0, group=None, dtype=torch.int64)
+        if isinstance(inputs, tuple):
+            context_tokens_tensor, context_length_tensor = inputs
+        else:
+            context_tokens_tensor, context_length_tensor = tokenize_batch(
+                inputs, self.tokenizer, self.cfg.encoder_seq_length, add_BOS=add_BOS, add_EOS=add_EOS,
+            )
 
-        # Select subset of data needed for this rank.
-        dp_size = parallel_state.get_data_parallel_world_size()
-        dp_rank = parallel_state.get_data_parallel_rank()
-        context_length_tensor = context_length_tensor.chunk(dp_size)[dp_rank]
-        context_tokens_tensor = context_tokens_tensor.chunk(dp_size)[dp_rank]
+        context_tokens_tensor = context_tokens_tensor.cuda()
+        context_length_tensor = context_length_tensor.cuda()
 
+        inference_batch_size, sequence_length = context_tokens_tensor.size()
+        forward_micro_batch_size = self.forward_micro_batch_size
         attention_mask, _, position_ids = get_ltor_masks_and_position_ids(
             context_tokens_tensor,
-            tokenizer.eos_id,
+            self.tokenizer.eos_id,
             self.cfg.get("reset_position_ids", False),
             self.cfg.get("reset_attention_mask", False),
             self.cfg.get("eod_mask_loss", False),
         )
-        micro_batch_size = context_tokens_tensor.shape[0]
-        sequence_length = context_tokens_tensor.shape[1]
-
-        app_state = AppState()
-        _reconfigure_microbatch_calculator(
-            rank=app_state.global_rank,
-            rampup_batch_size=None,
-            global_batch_size=micro_batch_size,
-            micro_batch_size=micro_batch_size,
-            data_parallel_size=1,
-        )
-        attention_mask_repeat = torch.concat([attention_mask for _ in range(micro_batch_size)])
-        rewards = self.forward_step(
-            [context_tokens_tensor, context_length_tensor, position_ids, attention_mask_repeat],
-            micro_batch_size,
-            sequence_length,
-        )
+        attention_mask = attention_mask.expand(inference_batch_size, -1, -1, -1)
+        inputs = [context_tokens_tensor, context_length_tensor, position_ids, attention_mask]
+
+        # if it's not divisible by forward micro batch size,
+        # then try our best to find a smaller one that is divisible
+        while inference_batch_size % forward_micro_batch_size != 0:
+            forward_micro_batch_size -= 1
```

Comment:
that's fair. The initial(perhaps wrong) thinking I had was that most of the time users will specify a preferred batch size to PyTriton, which means this case is rare. 

If this case was ever hit(for example if the user specifies a preferred batch size of a prime number), then they should modify their forward mbs accordingly. Otherwise, as long as their preferred batch size is a multiple of forward batch size then this code only acts as fall back if pytriton can't batch enough.

I do think this code is a bit ugly though... maybe it's better if we just pad it? I can do this easily, but that means if the user specifies `preferred_batch_size < forward micro batch size` then we no longer respect that, since we're padding

Reply:
What about:
* If `inference_batch_size < forward_micro_batch_size` => use a single batch of size `inference_batch_size`
* If `inference_batch_size > forward_micro_batch_size` => pad to multiple of `forward_micro_batch_size`

Reply:
yup! done here: 390eb1559c478b8033d6411144f422d17654b373

Reply:
also moved the server callable functions to server_utils since we don't use those callables 950e3d7004baf50b438ffe3f926fec44b5e1fc51

Reply:
Looks great, thanks!